crates/core/component/funding/src/component/liquidity_tournament/votes/mod.rs

```diff
@@ -2,43 +2,195 @@ use std::pin::Pin;
 use penumbra_sdk_asset::asset;
 use penumbra_sdk_keys::Address;
+use penumbra_sdk_sct::component::source::SourceContext as _;
+use penumbra_sdk_txhash::TransactionId;
 
 use crate::component::state_key;
 
-pub trait StateReadExt: StateRead {
-    fn vote_receipts(
+async fn get_u64(state: impl StateRead, key: &[u8]) -> u64 {
+    state
+        .nonverifiable_get_raw(key)
+        .await
+        .ok()
+        .and_then(|x| Some(u64::from_le_bytes(TryInto::<[u8; 8]>::try_into(x?).ok()?)))
+        .unwrap_or_default()
+}
+
+async fn put_u64(mut state: impl StateWrite, key: Vec<u8>, value: u64) {
+    state.nonverifiable_put_raw(key, value.to_le_bytes().to_vec());
+}
+
+async fn get_total(state: impl StateRead, epoch_index: u64) -> u64 {
+    get_u64(state, &state_key::lqt::v1::votes::total::key(epoch_index)).await
+}
+
+async fn put_total(state: impl StateWrite, epoch_index: u64, total: u64) {
+    put_u64(
+        state,
+        state_key::lqt::v1::votes::total::key(epoch_index).to_vec(),
+        total,
+    )
+    .await
+}
+
+async fn get_asset_total(state: impl StateRead, epoch_index: u64, asset: asset::Id) -> u64 {
+    get_u64(
+        state,
+        &state_key::lqt::v1::votes::by_asset::total::key(epoch_index, asset),
+    )
+    .await
+}
+
+async fn put_asset_total(state: impl StateWrite, epoch_index: u64, asset: asset::Id, total: u64) {
+    put_u64(
+        state,
+        state_key::lqt::v1::votes::by_asset::total::key(epoch_index, asset).to_vec(),
+        total,
+    )
+    .await
+}
+
+fn put_ranked_asset(mut state: impl StateWrite, epoch_index: u64, power: u64, asset: asset::Id) {
+    state.nonverifiable_put_raw(
+        state_key::lqt::v1::votes::by_asset::ranked::key(epoch_index, power, asset).to_vec(),
+        Vec::new(),
+    )
+}
+
+fn delete_ranked_asset(mut state: impl StateWrite, epoch_index: u64, power: u64, asset: asset::Id) {
+    state.nonverifiable_delete(
+        state_key::lqt::v1::votes::by_asset::ranked::key(epoch_index, power, asset).to_vec(),
+    );
+}
+
+async fn get_voter_total(
+    state: impl StateRead,
+    epoch_index: u64,
+    asset: asset::Id,
+    voter: &Address,
+) -> u64 {
+    get_u64(
+        state,
+        &state_key::lqt::v1::votes::by_voter::total::key(epoch_index, asset, voter),
+    )
+    .await
+}
+
+async fn put_voter_total(
+    state: impl StateWrite,
+    epoch_index: u64,
+    asset: asset::Id,
+    voter: &Address,
+    total: u64,
+) {
+    put_u64(
+        state,
+        state_key::lqt::v1::votes::by_voter::total::key(epoch_index, asset, voter).to_vec(),
+        total,
+    )
+    .await
+}
+
+fn put_ranked_voter(
+    mut state: impl StateWrite,
+    epoch_index: u64,
+    asset: asset::Id,
+    power: u64,
+    voter: &Address,
+    tx: TransactionId,
+) {
+    state.nonverifiable_put_raw(
+        state_key::lqt::v1::votes::by_voter::ranked::key(epoch_index, asset, power, voter).to_vec(),
+        tx.0.to_vec(),
```

Comment:
Same thing here, we use the proto overlay to get consistent codecs across the protocol